core/state/snapshot/snapshot.go

```diff
@@ -199,16 +200,16 @@ func New(config Config, diskdb ethdb.KeyValueStore, triedb *trie.Database, root
 		triedb: triedb,
 		layers: make(map[common.Hash]snapshot),
 	}
-	// Create the building waiter iff the background generation is allowed
-	if !config.NoBuild && !config.AsyncBuild {
-		defer snap.waitBuild()
-	}
 	// Attempt to load a previously persisted snapshot and rebuild one if failed
 	head, disabled, err := loadSnapshot(diskdb, triedb, root, config.CacheSize, config.Recovery, config.NoBuild)
 	if disabled {
 		log.Warn("Snapshot maintenance disabled (syncing)")
 		return snap, nil
 	}
+	// Create the building waiter iff the background generation is allowed
+	if !config.NoBuild && !config.AsyncBuild {
+		defer snap.waitBuild()
+	}
```

Comment:
This is an interesting change, and not immediately obvious why it was needed (or correct). It would make a difference in case we're syncing, and in the old code we would create a goroutine to `waitBuild`, and in the new code we would not. 

Any comments about this?

Reply:
Previously, we will always create a waiter if "sync-style" snapshot construction is required. Although in production, sync-style generation is always disabled which means the waiter is never be created.

While if we are in sync, the snapshot construction is disabled, if we still wait the construction by waiter, then it's a deadlook.

This change ensures we only create the waiter that the construction is really initialized.